icechunk/src/storage/s3.rs

```diff
@@ -29,28 +29,28 @@ use futures::{
     StreamExt, TryStreamExt,
 };
 use serde::{Deserialize, Serialize};
+use tokio::sync::RwLock;
 
 use crate::{
     format::{
         attributes::AttributesTable, format_constants, manifest::Manifest,
         snapshot::Snapshot, transaction_log::TransactionLog, AttributesId, ByteRange,
-        ChunkId, FileTypeTag, ManifestId, SnapshotId,
+        ChunkId, FileTypeTag, ManifestId, ObjectId, SnapshotId,
     },
-    private,
-    zarr::ObjectId,
-    Storage, StorageError,
+    private, Storage, StorageError,
 };
 
 use super::{
     ETag, ListInfo, StorageResult, CHUNK_PREFIX, CONFIG_PATH, MANIFEST_PREFIX,
     REF_PREFIX, SNAPSHOT_PREFIX, TRANSACTION_PREFIX,
 };
 
-#[derive(Debug)]
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(transparent)]
 pub struct S3Storage {
-    client: Arc<Client>,
-    prefix: String,
-    bucket: String,
+    config: S3Config,
+    #[serde(skip)]
+    client: RwLock<Option<Arc<Client>>>,
```

Comment:
Not a fan of the RwLock here. Why don't we use a OnceCell? That probably can replace both the RwLock and the Option.

Can you add a code comment explaining why we need the OnceCell to help readers? I understand it is because we don't serialize the client, so on deserialization it's initialized empty. But why can't we initialize it on deseralization and we skip the need for mutability completely?

Reply:
Done, this also lets me get rid of the `Arc`